dspy/clients/lm.py

```diff
@@ -151,3 +246,139 @@ def _inspect_history(lm, n: int = 1):
+
+from dspy.clients.openai import (
+    FinetuneJobOpenAI,
+    is_openai_model,
+    finetune_openai,
+)
+from dspy.clients.anyscale import (
+    FinetuneJobAnyScale,
+    finetune_anyscale,
+)
+
+
+_PROVIDER_ANYSCALE = "anyscale"
+_PROVIDER_OPENAI = "openai"
+_SUPPORTED_FINETUNE_PROVIDERS = [
+    _PROVIDER_ANYSCALE,
+    _PROVIDER_OPENAI,
+]
+
+            
+def _get_supported_finetune_provider(model: str) -> Union[str, ValueError]:
+    """Return the finetuning provider for the given model.
+
+    The provider must be in _SUPPORTED_FINETUNE_PROVIDERS. This function is not
+    named `_get_provider` because it does not attempt to find the provider of a
+    model if there is no DSPy fine-tuning support for it.
+    """
+    if is_openai_model(model):
+        return _PROVIDER_OPENAI
+
+    if is_anyscale_model(model):
+        return _PROVIDER_ANYSCALE
+    
+    return ValueError(f"DSPy does not have fine-tuning support for {model}")
+
+
+def get_provider_finetune_job_class(provider: str) -> Type[FinetuneJob]:
+    """Get the FinetuneJob class for the provider."""
+    # Mapping from provider to finetune job type
+    _PROVIDER_TO_FINETUNE_JOB_CLASS = {
+        _PROVIDER_ANYSCALE: FinetuneJobAnyScale,
+        _PROVIDER_OPENAI: FinetuneJobOpenAI,
+    }
+
+    # Get the FinetuneJob class for the provider
+    _CLS = _PROVIDER_TO_FINETUNE_JOB_CLASS[provider]
+
+    return _CLS
+
+
+def get_provider_finetune_function(provider: str) -> callable:
+    """Return the finetune function for the given model."""
+    # Mapping from provider to finetune function
+    _PROVIDER_TO_FINETUNE_FUNCTION = {
+        _PROVIDER_ANYSCALE: finetune_anyscale,
+        _PROVIDER_OPENAI: finetune_openai,
+    }
+
+    # Get the finetuning provider
+    finetune_function = _PROVIDER_TO_FINETUNE_FUNCTION[provider]
+
+    return finetune_function
+
+
+def execute_finetune_job(
+    job: FinetuneJob[Type[LM]],
+    lm: LM,
+    cache_finetune: bool=True
+):
+    """Execute the finetune job in a blocking manner."""
+    try:
+        job_kwargs = job.get_kwargs()
+        if cache_finetune:
+            model = cached_finetune(job=job, **job_kwargs)
+        else:
+            model = finetune(job=job, **job_kwargs)
+        lm = lm.copy(model=model)
+        job.set_result(lm)
+    except Exception as err:
+        logger.error(err)
+        job.set_result(err)
+
+
+# TODO: Perhaps we shouldn't directly cache the data
+# TODO: Add DiskCache, ignore job
+
+def cached_finetune(
+    job,
+    model: str,
+    train_data: List[Dict[str, Any]],
+    train_kwargs: Optional[Dict[str, Any]]=None,
+    provider: str = "openai",
+    train_method: TrainingMethod = TrainingMethod.SFT,
+) -> Union[str, ValueError]:
+    return finetune(
+        job=job,
+        model=model,
+        provider=provider,
+        train_data=train_data,
+        train_kwargs=train_kwargs,
+        train_method=train_method,
+    )
+
+
+def finetune(
```

Comment:
more boilerplate that shouldn't need to exist IMO

Reply:
Next PR again